File: src/main/java/org/distril/beengine/material/block/BlockState.java
package org.distril.beengine.material.block;

import com.nukkitx.nbt.NbtMap;
import lombok.AllArgsConstructor;
import lombok.Getter;
import org.distril.beengine.material.block.state.State;

@Getter
@AllArgsConstructor
public class BlockState {

	private NbtMap states;

	public <T> void set(State<T> state, T value) {
		var builder = this.states.getCompound("states").toBuilder();
		builder.put(state.getProperty(), value);
		this.states = this.states.toBuilder().putCompound("states", builder.build()).build();
	}

	@SuppressWarnings("unchecked")
	public <T> T get(State<T> state) {
		return (T) this.states.get(state.getProperty());
	}

	public int getRuntimeId() {
		return BlockPalette.getRuntimeId(this.states);
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj) {
			return true;
		}

		if (obj == null || this.getClass() != obj.getClass()) {
			return false;
		}

		BlockState that = (BlockState) obj;
		return this.states.equals(that.getStates());
	}

	@Override
	public int hashCode() {
		return this.states.hashCode();
	}
}

File: src/main/java/org/distril/beengine/material/block/BlockState.kt
package org.distril.beengine.material.block

import com.nukkitx.nbt.NbtMap
import org.distril.beengine.material.block.state.State

class BlockState(var states: NbtMap) {

	operator fun <T> set(state: State<T>, value: T) {
		val builder = this.states.getCompound("states").toBuilder()
		builder[state.property] = value
		this.states = states.toBuilder().putCompound("states", builder.build()).build()
	}

	@Suppress("UNCHECKED_CAST")
	operator fun <T> get(state: State<T>) = this.states[state.property] as T

	val runtimeId = BlockPalette.getRuntimeId(this.states)

	override fun equals(other: Any?): Boolean {
		if (this === other) return true
		if (other !is BlockState) return false

		return this.states == other.states
	}

	override fun hashCode() = this.states.hashCode()
}
